File: opencga-client/src/main/python/pyOpenCGA/pyopencga/rest_clients/clinical_client.py
from pyopencga.rest_clients._parent_rest_clients import _ParentRestClient, _ParentBasicCRUDClient, _ParentAclRestClient
from pyopencga.commons import deprecated

class Interpretations(_ParentRestClient, _ParentAclRestClient):
    """
    This class contains the Interpretations client with methods for the
    Analysis - Clinical webservices
    """

    def __init__(self, configuration, token=None, login_handler=None, *args, **kwargs):
        _category = 'analysis/clinical'
        super(Interpretations, self).__init__(configuration, _category, token, login_handler, *args, **kwargs)

    @deprecated
    def tool_tiering(self, **options):
        return self.run_tiering(**options)

    def run_tiering(self, **options):
        """
        GEL Tiering interpretation analysis
        URL: /{apiVersion}/analysis/clinical/interpretation/tiering/run
        """

        return self._post('interpretation/tiering/run', **options)

    @deprecated
    def tool_team(self, **options):
        return self.run_team(**options)

    def run_team(self, **options):
        """
        TEAM interpretation analysis
        URL: /{apiVersion}/analysis/clinical/interpretation/team/run
        """

        return self._post('interpretation/team/run', **options)

    @deprecated
    def tool_custom(self, **options):
        return self.run_custom(**options)

    def run_custom(self, **options):
        """
        TEAM interpretation analysis
        URL: /{apiVersion}/analysis/clinical/interpretation/custom/run
        """

        return self._post('interpretation/custom/run', **options)

    def run_cancer_tiering(self, **options):
        """
        Cancer Tiering interpretation analysis
        URL: /{apiVersion}/analysis/clinical/interpretation/cancerTiering/run
        """

        return self._post('interpretation/cancerTiering/run', **options)

    def primary_findings(self, **options):
        """
        Search for secondary findings for a given query
        URL: /{apiVersion}/analysis/clinical/interpretation/primaryFindings
        """

        return self._get('interpretation/primaryFindings', **options)

    def secondary_findings(self, **options):
        """
        Search for secondary findings for a given sample
        URL: /{apiVersion}/analysis/clinical/interpretation/secondaryFindings
        """

        return self._get('interpretation/secondaryFindings', **options)

    def stats(self, **options):
        """
        Clinical interpretation analysis
        URL: /{apiVersion}/analysis/clinical/interpretation/stats
        """

        return self._get('interpretation', subcategory='stats', **options)

    def query(self, **options):
        """
        Query for reported variants
        URL: /{apiVersion}/analysis/clinical/interpretation/query
        """

        return self._get('interpretation', subcategory='query', **options)

    def index(self, **options):
        """
        Index clinical analysis interpretations in the clinical variant database
        URL: /{apiVersion}/analysis/clinical/interpretation/index
        """

        return self._get('interpretation', subcategory='index', **options)

    def update_interpretation(self, clinical_analysis, data, action, **options):
        """
        Add or remove Interpretations to/from a Clinical Analysis
        URL: /{apiVersion}/analysis/clinical/{clinicalAnalysis}/interpretations/update

        :param clinical_analysis: clinical analysis id
        :param action: Action to be performed if the array of interpretations is being updated [ADD, REMOVE]
        """

        return self._post('interpretations', query_id=clinical_analysis, subcategory='update', data=data, action=action, **options)

    def update_interpretation_data(self, clinical_analysis, interpretation, data, **options):
        """
        Update Interpretation fields
        URL: /{apiVersion}/analysis/clinical/{clinicalAnalysis}/interpretations/{interpretation}/update

        :param clinical_analysis: clinical analysis id
        :param interpretation: interpretation id
        """

        return self._post('interpretations', query_id=clinical_analysis, subcategory='update',
                          subquery_id=interpretation, data=data, **options)

    def update_interpretation_comments(self, clinical_analysis, interpretation, data, **options):
        """
        Update comments of an Interpretation
        URL: /{apiVersion}/analysis/clinical/{clinicalAnalysis}/interpretations/{interpretation}/comment/update

        :param clinical_analysis: clinical analysis id
        :param interpretation: interpretation id
        """

        return self._post('interpretations', query_id=clinical_analysis, subcategory='comment/update',
                          subquery_id=interpretation, data=data, **options)

    def update_interpretation_primary_findings(self, clinical_analysis, interpretation, data, **options):
        """
        Update reported variants of an interpretation
        URL: /{apiVersion}/analysis/clinical/{clinicalAnalysis}/interpretations/{interpretation}/primaryFindings/update

        :param clinical_analysis: clinical analysis id
        :param interpretation: interpretation id
        """

        return self._post('interpretations', query_id=clinical_analysis, subcategory='primaryFindings/update',
                          subquery_id=interpretation, data=data, **options)

    def update_comments(self, clinical_analysis, interpretation, data, action, **options):
        """
        Update comments of an Interpretation
        URL: /{apiVersion}/analysis/clinical/{clinicalAnalysis}/interpretations/{interpretation}/comments/update

        :param action: Action to be performed [ADD, SET, REMOVE]
        """

        return self._post('interpretations', query_id=clinical_analysis, subquery_id=interpretation,
                          subcategory='comments/update', data=data, action=action, **options)

    def update_reported_variants(self, clinical_analysis, interpretation, data, **options):
        """
        Update reported variants of an interpretation
        URL: /{apiVersion}/analysis/clinical/{clinicalAnalysis}/interpretations/{interpretation}/reportedVariants/update
        """

        return self._post('interpretations', query_id=clinical_analysis, subquery_id=interpretation,
                           subcategory='reportedVariants/update', data=data, **options)


class Clinical(_ParentBasicCRUDClient, _ParentAclRestClient):
    """
    This class contains methods for the Analysis - Clinical webservices
    """

    def __init__(self, configuration, token=None, login_handler=None, *args, **kwargs):
        _category  = 'analysis/clinical'
        super(Clinical, self).__init__(configuration, _category, token, login_handler, *args, **kwargs)

        self.configuration = configuration
        self.token = token
        self.login_handler = login_handler
        self._create_clients()

    def _create_clients(self):

        ## undef all
        self.interpretations = None
        self.reports = None
        self.cva = None

        ## [TODO] convert to @properties
        ## [@dgp] SHould I add auto_refresh = self.auto_refresh ??
        self.interpretations = Interpretations(self.configuration, self.token, self.login_handler)
        ## self.reports = Reports(configuration, token, login_handler, *args, **kwargs)
        ## self.cva = CVAs(configuration, token, login_handler, *args, **kwargs)

        self.clients = [self.interpretations]

        for client in self.clients:
            # only retry the ones with objects (instatiated clients)
            if client is not None:
                client.on_retry = self.on_retry

    def search(self, **options):
        """
        Clinical analysis search.
        URL: /{apiVersion}/analysis/clinical/search
        """

        if 'skipCount' not in options and ('count' not in options or options['count'] is False):
            options['skipCount'] = True

        return self._get('search', **options)

File: opencga-client/src/main/java/org/opencb/opencga/client/rest/ClinicalClient.java
/*
* Copyright 2015-2020 OpenCB
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

package org.opencb.opencga.client.rest;

import java.util.List;
import org.opencb.biodata.models.clinical.interpretation.Interpretation;
import org.opencb.biodata.models.clinical.interpretation.ReportedVariant;
import org.opencb.commons.datastore.core.ObjectMap;
import org.opencb.opencga.client.config.ClientConfiguration;
import org.opencb.opencga.client.exceptions.ClientException;
import org.opencb.opencga.core.models.clinical.ClinicalAnalysis;
import org.opencb.opencga.core.models.clinical.ClinicalAnalysisAclUpdateParams;
import org.opencb.opencga.core.models.clinical.ClinicalAnalysisCreateParams;
import org.opencb.opencga.core.models.clinical.ClinicalUpdateParams;
import org.opencb.opencga.core.models.clinical.InterpretationCreateParams;
import org.opencb.opencga.core.models.clinical.InterpretationUpdateParams;
import org.opencb.opencga.core.models.job.Job;
import org.opencb.opencga.core.response.RestResponse;


/**
 * This class contains methods for the Clinical webservices.
 *    Client version: 2.0.0
 *    PATH: analysis/clinical
 */
public class ClinicalClient extends AbstractParentClient {

    public ClinicalClient(String token, ClientConfiguration configuration) {
        super(token, configuration);
    }

    /**
     * Index clinical analysis interpretations in the clinical variant database.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<ObjectMap> indexInterpretation(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/clinical", null, "interpretation", null, "index", params, GET, ObjectMap.class);
    }

    /**
     * Update clinical analysis attributes.
     * @param clinicalAnalyses Comma separated list of clinical analysis ids.
     * @param data JSON containing clinical analysis information.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<ClinicalAnalysis> update(String clinicalAnalyses, ClinicalUpdateParams data, ObjectMap params)
            throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.put("body", data);
        return execute("analysis/clinical", clinicalAnalyses, null, null, "update", params, POST, ClinicalAnalysis.class);
    }

    /**
     * Update Interpretation fields.
     * @param clinicalAnalysis Clinical analysis id.
     * @param interpretation Interpretation id.
     * @param data JSON containing clinical interpretation information.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Interpretation> updateInterpretation(String clinicalAnalysis, String interpretation, InterpretationUpdateParams
        data, ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.put("body", data);
        return execute("analysis/clinical", clinicalAnalysis, "interpretations", interpretation, "update", params, POST,
                Interpretation.class);
    }

    /**
     * Create a new clinical analysis.
     * @param data JSON containing clinical analysis information.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<ClinicalAnalysis> create(ClinicalAnalysisCreateParams data, ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.put("body", data);
        return execute("analysis/clinical", null, null, null, "create", params, POST, ClinicalAnalysis.class);
    }

    /**
     * Query for reported variants.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<ObjectMap> queryInterpretation(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/clinical", null, "interpretation", null, "query", params, GET, ObjectMap.class);
    }

    /**
     * Clinical analysis search.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<ClinicalAnalysis> search(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/clinical", null, null, null, "search", params, GET, ClinicalAnalysis.class);
    }

    /**
     * Clinical interpretation analysis.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<ObjectMap> statsInterpretation(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/clinical", null, "interpretation", null, "stats", params, GET, ObjectMap.class);
    }

    /**
     * Clinical analysis info.
     * @param clinicalAnalyses Comma separated list of clinical analysis IDs or names up to a maximum of 100.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<ClinicalAnalysis> info(String clinicalAnalyses, ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/clinical", clinicalAnalyses, null, null, "info", params, GET, ClinicalAnalysis.class);
    }

    /**
     * Returns the acl of the clinical analyses. If member is provided, it will only return the acl for the member.
     * @param clinicalAnalyses Comma separated list of clinical analysis IDs or names up to a maximum of 100.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<ObjectMap> acl(String clinicalAnalyses, ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/clinical", clinicalAnalyses, null, null, "acl", params, GET, ObjectMap.class);
    }

    /**
     * Update the set of permissions granted for the member.
     * @param members Comma separated list of user or group ids.
     * @param data JSON containing the parameters to add ACLs.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<ObjectMap> updateAcl(String members, ClinicalAnalysisAclUpdateParams data, ObjectMap params)
            throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.put("body", data);
        return execute("analysis/clinical", members, null, null, "update", params, POST, ObjectMap.class);
    }

    /**
     * Add or remove Interpretations to/from a Clinical Analysis.
     * @param clinicalAnalysis Clinical analysis ID.
     * @param data JSON containing clinical analysis information.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<ClinicalAnalysis> updateInterpretations(String clinicalAnalysis, InterpretationCreateParams data, ObjectMap params)
            throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.put("body", data);
        return execute("analysis/clinical", clinicalAnalysis, "interpretations", null, "update", params, POST, ClinicalAnalysis.class);
    }

    /**
     * Update comments of an Interpretation.
     * @param clinicalAnalysis Clinical analysis id.
     * @param interpretation Interpretation id.
     * @param data JSON containing a list of comments.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Interpretation> updateComments(String clinicalAnalysis, String interpretation, List data, ObjectMap params)
            throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.put("body", data);
        return execute("analysis/clinical", clinicalAnalysis, "interpretations", interpretation, "comments/update", params, POST,
                Interpretation.class);
    }

    /**
     * Update reported variants of an interpretation.
     * @param clinicalAnalysis Clinical analysis id.
     * @param interpretation Interpretation id.
     * @param data JSON containing a list of reported variants.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Interpretation> updatePrimaryFindings(String clinicalAnalysis, String interpretation, List data, ObjectMap params)
            throws ClientException {
        params = params != null ? params : new ObjectMap();
        params.put("body", data);
        return execute("analysis/clinical", clinicalAnalysis, "interpretations", interpretation, "primaryFindings/update", params, POST,
                Interpretation.class);
    }

    /**
     * TEAM interpretation analysis.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Job> runInterpretationTeam(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/clinical", null, "interpretation/team", null, "run", params, POST, Job.class);
    }

    /**
     * GEL Tiering interpretation analysis.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Job> runInterpretationTiering(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/clinical", null, "interpretation/tiering", null, "run", params, POST, Job.class);
    }

    /**
     * Interpretation custom analysis.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Job> runInterpretationCustom(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/clinical", null, "interpretation/custom", null, "run", params, POST, Job.class);
    }

    /**
     * Cancer Tiering interpretation analysis.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<Job> runInterpretationCancerTiering(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/clinical", null, "interpretation/cancerTiering", null, "run", params, POST, Job.class);
    }

    /**
     * Search for secondary findings for a given query.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<ReportedVariant> primaryFindingsInterpretation(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/clinical", null, "interpretation", null, "primaryFindings", params, GET, ReportedVariant.class);
    }

    /**
     * Search for secondary findings for a given sample.
     * @param params Map containing any additional optional parameters.
     * @return a RestResponse object.
     * @throws ClientException ClientException if there is any server error.
     */
    public RestResponse<ReportedVariant> secondaryFindingsInterpretation(ObjectMap params) throws ClientException {
        params = params != null ? params : new ObjectMap();
        return execute("analysis/clinical", null, "interpretation", null, "secondaryFindings", params, GET, ReportedVariant.class);
    }
}
